js/stable.js

```diff
@@ -1130,61 +1130,60 @@ dxSTable.prototype.addRow = function(cols, sId, icon, attr)
 	return true;
 }
 
-dxSTable.prototype.createIconHTML = function(icon)
-{
-	return icon == null ? '' : (typeof icon === 'object'
-		? $('<img>').attr({src: icon.src, width: 16, height: 16}).css('background-image', 'none').addClass('stable-icon')
-		: $('<span>').addClass(['stable-icon', icon]))[0].outerHTML;
+dxSTable.prototype.createIconHTML = function(icon) {
+	// TODO: deprecated, remove this method in v6
+	noty("Deprecation warning: `dxSTable.createIconHTML()` is deprecated. Please use `dxSTable.createIcon()` instead.");
+	return this.createIcon(icon)[0].outerHTML;
+}
+
+dxSTable.prototype.createIcon = function(icon) {
+	if (!icon) return "";
+	const iconObj = $("<span>").addClass("stable-icon");
+	if ($type(icon) === "object") {
+		return iconObj.css("background-image", `url(${icon.src})`);
+	} else {
+		return iconObj.addClass(icon).css("background-image", "");
+	}
 }
 
-dxSTable.prototype.createRow = function(cols, sId, icon, attr)
-{
+dxSTable.prototype.createRow = function(cols, sId, icon, attr) {
 	const attrs = { id: sId, index: this.rows, title: cols[0] };
 	if (sId == null) {
 		delete attrs['id'];
 	}
 	Object.assign(attrs, attr || {});
 	const data = this.rowdata[sId]?.fmtdata || {};
 
-	const ret = document.createElement('tr');
-	ret.className = this.colorEvenRows ? ((this.rows & 1) ? "odd" : "even") : "";
-	for(const [a,v] of Object.entries(attrs)) {
-		const attr_node = document.createAttribute(a);
-		attr_node.value = v;
-		ret.setAttributeNode(attr_node);
+	const row = $("<tr>").attr(attrs).addClass(this.colorEvenRows ? ((this.rows & 1) ? "odd" : "even") : "");
+	for (let i = 0; i < this.cols; i++) {
+		const index = this.colOrder[i];
+		const cdat = this.colsdata[i];
+		const td = $("<td>").addClass(`stable-${this.dCont.id}-col-${index}`).toggle(!!(cdat.enabled || browser.isIE7x));
+		const celldata = data[index] || '';
+		const rawvalue = cols[index] || '';
+		const isProgress = cdat.type === TYPE_PROGRESS;
+		if (isProgress) {
+			td.attr({rawvalue:rawvalue}).append(
+				$("<span>").addClass("meter-text").css({overflow:"visible"}).text(celldata),
+				$("<div>")
+					.addClass("meter-value")
+					.css(this.progressStyle(celldata)),
+			);
+		} else {
+			td.append(
+				$("<div>").text(celldata || " "),
+			);
+		}
+		row.append(td);
 	}
-	ret.innerHTML = [...Array(this.cols).keys()]
-			.map((_,i) => [this.colOrder[i], this.colsdata[i]])
-			.map(([ind, cdat]) => ({
-				td: [
-					`<td class="stable-${this.dCont.id}-col-${ind}"`,
-					Boolean(cdat.enabled || browser.isIE7x) ?	'>' : ' style="display: none">',
-					ind === 0 ? this.createIconHTML(icon) : ''
```

Comment:
This conditional runs on every cell. If there are 25 cells in a row, and 1,000 rows in a table, there will be 24 * 1,000 = 24,000 ineffective checks.